### PR TITLE
TLS 1.3: EarlyData SRV: Add early data extension parser.

### LINKED ISSUE #6331: TLS 1.3 server: Parsing of the early data indication extension
This issue tracks the up-streaming on server side of the parsing of the early data indication extension

Source: https://github.com/hannestschofenig/mbedtls/tree/tls13-prototype

PLEASE READ THOROUGHLY section 4.2.10 of the specification

- [x] Parse the early data indication extension in the ClientHello and note down its reception in `ssl->handshake->extensions_present`. 
- [ ] Abort the handshake if an early data indication extension is received in a second ClientHello (in response to the server sending an HRR)

Testing:
test (ssl-opt.sh test) with a GnuTLS client (not possible with an OpenSSL client I think) that a "resumption" ClientHello containing an early data indication extension is successfully parsed by the MbedTLS server. The resumption handshake is expected to fail at this stage.

Depends on #6330

### PR DESCRIPTION
**Pull request by yuhaoth**
## Description

Fix  #6331

- [x] When `max_early_data_size` is disabled and received early data extension, ignore `decrypt error`
- [x] Relative tests.

## Gatekeeper checklist

- [x] **changelog** provided, or not required
- [x] **backport** done, or not required
- [x] **tests** provided, or not required

## Notes for the submitter

Please refer to the [contributing guidelines](../CONTRIBUTING.md), especially the
checklist for PR contributors.

### CODE REVIEW

**Comment by yuhaoth** [comment]
Internal CI passed at https://jenkins-mbedtls.oss.arm.com/blue/organizations/jenkins/mbed-tls-pr-merge/detail/PR-6719-merge/15/pipeline

Open CI passed at https://ci.staging.trustedfirmware.org/blue/organizations/jenkins/mbed-tls-pr-merge/detail/PR-6719-merge/15/pipeline

The latest tests were cancelled by CI.

File: library/ssl_tls13_server.c [hunk 1]
```diff
@@ -1701,6 +1749,17 @@ static int ssl_tls13_postprocess_client_hello( mbedtls_ssl_context* ssl )
         return( ret );
     }
 
+#if defined(MBEDTLS_SSL_EARLY_DATA)
+    /* There is enough information, update early data state. */
+    ssl_tls13_update_early_data_state( ssl );
+#elif defined(MBEDTLS_DEBUG_C)
+    if( !mbedtls_ssl_tls13_early_data_is_rejected( ssl ) &&
+        mbedtls_ssl_tls13_early_data_is_received( ssl ) )
```
Comment (lpy4105) [inline library/ssl_tls13_server.c:1757]: Not sure what case you mean to check here, but currently if `early_data_state` is set to `xxx_ACCEPTED`, `"should never happend."` would be printed.

Comment (yuhaoth) [reply library/ssl_tls13_server.c:1757]: It has been removed in my local version. will update it.

**Review by mprse — COMMENTED** [review]
PR looks good. Found few potential places for improvements or clarification.

File: library/ssl_tls13_server.c [hunk 2]
```diff
@@ -1673,6 +1673,54 @@ static int ssl_tls13_parse_client_hello( mbedtls_ssl_context *ssl,
...
+                                            MBEDTLS_SSL_MAX_EARLY_DATA_SIZE;
+    }
+
+    MBEDTLS_SSL_DEBUG_MSG(
+            3, ( "EarlyData: conf->max_early_data_size = %u",
+                 (unsigned int)ssl->conf->max_early_data_size ) );
+
+    /* TODO: Add more checks here. */
+
+    MBEDTLS_SSL_DEBUG_MSG(
```
[the remark this answers is not in the document]
Comment (yuhaoth) [reply library/ssl_tls13_server.c:1717]: Yes. In #6720 and #6721, more checks is added

File: library/ssl_tls13_server.c [hunk 3]
```diff
@@ -1673,6 +1673,54 @@ static int ssl_tls13_parse_client_hello( mbedtls_ssl_context *ssl,
     return( hrr_required ? SSL_CLIENT_HELLO_HRR_REQUIRED : SSL_CLIENT_HELLO_OK );
 }
 
+#if defined(MBEDTLS_SSL_EARLY_DATA)
+static void ssl_tls13_update_early_data_state( mbedtls_ssl_context *ssl )
+{
+    mbedtls_ssl_session *session = ssl->session_negotiate;
+    mbedtls_ssl_handshake_params *handshake = ssl->handshake;
+    ((void) session);
```
[the remark this answers is not in the document]
Comment (yuhaoth) [reply library/ssl_tls13_server.c:1681]: Yes. It is added in #6720.

File: library/ssl_tls13_server.c [hunk 4]
```diff
@@ -1673,6 +1673,54 @@ static int ssl_tls13_parse_client_hello( mbedtls_ssl_context *ssl,
...
+                 (unsigned int)MBEDTLS_SSL_MAX_EARLY_DATA_SIZE ) );
+        ((mbedtls_ssl_config *) ssl->conf)->max_early_data_size =
+                                            MBEDTLS_SSL_MAX_EARLY_DATA_SIZE;
+    }
+
+    MBEDTLS_SSL_DEBUG_MSG(
+            3, ( "EarlyData: conf->max_early_data_size = %u",
+                 (unsigned int)ssl->conf->max_early_data_size ) );
+
+    /* TODO: Add more checks here. */
```
[the remark this answers is not in the document]
Comment (yuhaoth) [reply library/ssl_tls13_server.c:1715]: session related checks is added in #6720. And I forget move session variable to #6720

File: library/ssl_misc.h [hunk 5]
```diff
@@ -2067,6 +2080,45 @@ int mbedtls_ssl_tls13_write_early_data_ext( mbedtls_ssl_context *ssl,
...
                                             size_t *out_len );
+
+#if defined(MBEDTLS_SSL_SRV_C)
+#define MBEDTLS_SSL_EARLY_DATA_SRV_STATE_NOT_RECEIVED    0
+#define MBEDTLS_SSL_EARLY_DATA_SRV_STATE_REJECTED        1
+#define MBEDTLS_SSL_EARLY_DATA_SRV_STATE_ACCEPTED        2
+static inline uint8_t mbedtls_ssl_tls13_early_data_state(
+                          mbedtls_ssl_context *ssl )
+{
+    if( ssl->handshake == NULL                        ||
```
[the remark this answers is not in the document]
Comment (yuhaoth) [reply library/ssl_misc.h:2091]: Thanks @mprse.

**Review by mprse — DISMISSED** [review]
Thanks for addressing my my comments.
LGTM

File: tests/ssl-opt.sh [hunk 6]
```diff
@@ -285,9 +286,8 @@ get_config_value_or_default() {
 requires_config_value_at_least() {
     VAL="$( get_config_value_or_default "$1" )"
     if [ -z "$VAL" ]; then
-        # Should never happen
-        echo "Mbed TLS configuration $1 is not defined"
-        exit 1
+        # when test-ref-config, MBEDTLS_SSL_MAX_EARLY_DATA_SIZE report undefine.
+        SKIP_NEXT="YES"
```
Comment (lpy4105) [inline tests/ssl-opt.sh:290]: Should we consider add "MBEDTLS_SSL_MAX_EARLY_DATA_SIZE" to `requires_all_configs_enabled` instead of change the behavior here?

**Review by lpy4105 — COMMENTED** [review]
LGTM, but with a minor question.

File: tests/ssl-opt.sh (hunk 6 above)
Comment (yuhaoth) [reply tests/ssl-opt.sh:290]: The issue is reported by shell script.
That won't resolve the script issue.  The sentence will be  called always.

**Comment by yuhaoth** [comment]
Just fix CI fail

File: library/ssl_misc.h [hunk 7]
```diff
@@ -2088,6 +2088,8 @@ int mbedtls_ssl_tls13_write_early_data_ext( mbedtls_ssl_context *ssl,
 static inline uint8_t mbedtls_ssl_tls13_early_data_state(
                           mbedtls_ssl_context *ssl )
 {
+    if( ssl->handshake == NULL )
```
Comment (mprse) [inline library/ssl_misc.h:2091]: So it seems that my analysis was wrong.

Comment (yuhaoth) [reply library/ssl_misc.h:2091]: That's my mistake. I should keep handshake check.

**Review by lpy4105 — DISMISSED** [review]
LGTM

**Comment by yuhaoth** [comment]
rebase to re-trigger CI test and switch coding style

**Comment by yuhaoth** [comment]
OpenCI reports PASS . But Internal CI reports node exception.

**Review by lpy4105 — COMMENTED** [review]
Some minor problems, others LGTM

**Review by lpy4105 — APPROVED** [review]
LGTM

File: library/ssl_msg.c [hunk 8]
```diff
@@ -3745,6 +3745,45 @@ static int ssl_prepare_record_content(mbedtls_ssl_context *ssl,
                                            rec)) != 0) {
             MBEDTLS_SSL_DEBUG_RET(1, "ssl_decrypt_buf", ret);
 
+#if defined(MBEDTLS_SSL_EARLY_DATA) && defined(MBEDTLS_SSL_SRV_C)
```
Comment (ronald-cron-arm) [inline library/ssl_msg.c:3748]: This is addressing #6360. It is better practice to address only one issue per PR (here #6331) and as the issue #6360 is scheduled for Q2 now, I'd prefer to move this to another PR and keep in this PR only the code addressing #6331.

File: library/ssl_misc.h [hunk 9]
```diff
@@ -985,11 +985,24 @@ struct mbedtls_ssl_handshake_params {
...
-#endif
+
+#if defined(MBEDTLS_SSL_SRV_C)
+    /** The state of server side early data  during handshake.
+     *  Available values are
+     *  - MBEDTLS_SSL_EARLY_DATA_SRV_STATE_NOT_RECEIVED
+     *  - MBEDTLS_SSL_EARLY_DATA_SRV_STATE_REJECTED
+     *  - MBEDTLS_SSL_EARLY_DATA_SRV_STATE_ACCEPTED
+     */
+    uint8_t early_data_state;
```
Comment (ronald-cron-arm) [inline library/ssl_misc.h:1001]: This is identical (same semantic) to `early_data_status` on client side. Thus why not using the same context field (`early_data_status` in `mbedtls_ssl_context`) and the same macros MBEDTLS_SSL_EARLY_DATA_STATUS_*. We probably do not want to expose it through the `mbedtls_ssl_get_early_data_status` API on server side though.

File: library/ssl_tls13_server.c [hunk 10]
```diff
@@ -1633,6 +1633,49 @@ static int ssl_tls13_parse_client_hello(mbedtls_ssl_context *ssl,
...
+
+    handshake->early_data_state = MBEDTLS_SSL_EARLY_DATA_SRV_STATE_REJECTED;
+
+    if (ssl->conf->early_data_enabled == MBEDTLS_SSL_EARLY_DATA_DISABLED) {
+        MBEDTLS_SSL_DEBUG_MSG(
+            1, ("EarlyData: rejected. configured disabled."));
+        return;
+    }
+
+    if (ssl->conf->max_early_data_size > MBEDTLS_SSL_MAX_EARLY_DATA_SIZE) {
```
Comment (ronald-cron-arm) [inline library/ssl_tls13_server.c:1656]: I do not understand why you do this. In mbedtls_ssl_read_early_data() we plan to store early data into buffers provided by the application thus I do not think our TLS 1.3 implementation should impose a compile time limit: MBEDTLS_SSL_MAX_EARLY_DATA_SIZE here.

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]

File: library/ssl_misc.h (hunk 9 above)
Comment (yuhaoth) [reply library/ssl_misc.h:1001]: Forget that. I plan to share it but forget that after `ssl->early_data_status` got merged.

File: library/ssl_tls13_server.c (hunk 10 above)
Comment (yuhaoth) [reply library/ssl_tls13_server.c:1656]: I think your are right.  will remove that

File: library/ssl_msg.c (hunk 8 above)
Comment (yuhaoth) [reply library/ssl_msg.c:3748]: I will move this commit to other PR

File: library/ssl_misc.h [hunk 11]
```diff
@@ -2095,6 +2097,41 @@ int mbedtls_ssl_tls13_write_early_data_ext(mbedtls_ssl_context *ssl,
                                            unsigned char *buf,
                                            const unsigned char *end,
                                            size_t *out_len);
+
+#if defined(MBEDTLS_SSL_SRV_C)
```
Comment (ronald-cron-arm) [inline library/ssl_misc.h:2101]: I do not think we need these utility functions, at least currently. It is defining new symbols that do not bring much than just accessing `early_data_status` directly.

File: library/ssl_tls13_server.c [hunk 12]
```diff
@@ -1633,6 +1633,39 @@ static int ssl_tls13_parse_client_hello(mbedtls_ssl_context *ssl,
     return hrr_required ? SSL_CLIENT_HELLO_HRR_REQUIRED : SSL_CLIENT_HELLO_OK;
 }
 
+#if defined(MBEDTLS_SSL_EARLY_DATA)
+static void ssl_tls13_update_early_data_state(mbedtls_ssl_context *ssl)
```
Comment (ronald-cron-arm) [inline library/ssl_tls13_server.c:1637]:
```suggestion
static void ssl_tls13_update_early_data_status(mbedtls_ssl_context *ssl)
```

File: programs/ssl/ssl_server2.c [hunk 13]
```diff
@@ -130,7 +130,7 @@ int main(void)
 #define DFL_SNI                 NULL
 #define DFL_ALPN_STRING         NULL
 #define DFL_CURVES              NULL
-#define DFL_MAX_EARLY_DATA_SIZE 0
+#define DFL_MAX_EARLY_DATA_SIZE NULL
```
Comment (ronald-cron-arm) [inline programs/ssl/ssl_server2.c:133]: The changes in this file are out of scope for this PR now. Please move them to another PR. And I would say it is the case for all the changes related to testing. Otherwise, instead of reverting more commits, as this PR is going to be quite small eventually, it is probably better to just to rewrite the history.

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]

File: programs/ssl/ssl_server2.c (hunk 13 above)
Comment (yuhaoth) [reply programs/ssl/ssl_server2.c:133]: `opt.max_early_data_size` is a string. So it should be NULL. 

I will move it to separate PR.

Comment (yuhaoth) [reply programs/ssl/ssl_server2.c]: I have rewritten the history.

**Comment by yuhaoth** [comment]
Internal CI pass.

File: library/ssl_misc.h [hunk 14]
```diff
@@ -2095,6 +2095,12 @@ int mbedtls_ssl_tls13_write_early_data_ext(mbedtls_ssl_context *ssl,
...
+#define MBEDTLS_SSL_EARLY_DATA_STATUS_NOT_RECEIVED \
+    MBEDTLS_SSL_EARLY_DATA_STATUS_NOT_SENT
+#endif /* MBEDTLS_SSL_SRV_C */
```
Comment (lpy4105) [inline library/ssl_misc.h:2102]: Should this be moved to `include/mbedtls/ssl.h`?

Comment (yuhaoth) [reply library/ssl_misc.h:2102]: Not, it should not be expose to user.

**Review by lpy4105 — APPROVED** [review]
LGTM

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]

File: tests/opt-testcases/tls13-misc.sh [hunk 15]
```diff
@@ -493,3 +493,35 @@ run_test "TLS 1.3 m->m: Resumption with ticket flags, psk_all/psk_all." \
          -S "No suitable key exchange mode" \
          -s "found matched identity"
 
+requires_gnutls_next
+requires_all_configs_enabled MBEDTLS_SSL_EARLY_DATA MBEDTLS_SSL_SESSION_TICKETS     \
+                             MBEDTLS_SSL_SRV_C MBEDTLS_DEBUG_C MBEDTLS_HAVE_TIME    \
+                             MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED \
+                             MBEDTLS_SSL_TLS1_3_COMPATIBILITY_MODE MBEDTLS_ECP_LIGHT
```
[the remark this answers is not in the document]
Comment (yuhaoth) [reply tests/opt-testcases/tls13-misc.sh:500]: removed

File: tests/opt-testcases/tls13-misc.sh [hunk 16]
```diff
@@ -493,3 +493,35 @@ run_test "TLS 1.3 m->m: Resumption with ticket flags, psk_all/psk_all." \
...
+         -S "EncryptedExtensions: early_data(42) extension exists."         \
+         -s "NewSessionTicket: early_data(42) extension does not exist."
+
+requires_gnutls_next
+requires_all_configs_enabled MBEDTLS_SSL_EARLY_DATA MBEDTLS_SSL_SESSION_TICKETS     \
+                             MBEDTLS_SSL_SRV_C MBEDTLS_DEBUG_C MBEDTLS_HAVE_TIME    \
+                             MBEDTLS_SSL_TLS1_3_COMPATIBILITY_MODE MBEDTLS_ECP_LIGHT
+requires_any_configs_enabled MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_ENABLED \
+                             MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_EPHEMERAL_ENABLED
+run_test "TLS 1.3 G->m: EarlyData: psk*: feature is disabled, fail." \
```
[the remark this answers is not in the document]
Comment (yuhaoth) [reply tests/opt-testcases/tls13-misc.sh:518]: Sure.

File: tests/opt-testcases/tls13-misc.sh [hunk 17]
```diff
@@ -493,3 +493,18 @@ run_test "TLS 1.3 m->m: Resumption with ticket flags, psk_all/psk_all." \
...
+                             MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED \
+                             MBEDTLS_SSL_TLS1_3_COMPATIBILITY_MODE
+requires_any_configs_enabled MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_ENABLED \
+                             MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_EPHEMERAL_ENABLED
+run_test "TLS 1.3 G->m: EarlyData: feature is disabled, fail." \
+         "$P_SRV force_version=tls13 debug_level=4 max_early_data_size=-1 $(get_srv_psk_list)" \
+         "$G_NEXT_CLI localhost --priority=NORMAL:-VERS-ALL:+VERS-TLS1.3:+GROUP-ALL -d 10 -r --earlydata $EARLY_DATA_INPUT" \
+         1 \
+         -s "ClientHello: early_data(42) extension exists."                     \
+         -s "EncryptedExtensions: early_data(42) extension does not exists."    \
```
Comment (ronald-cron-arm) [inline tests/opt-testcases/tls13-misc.sh:508]:
```suggestion
         -s "EncryptedExtensions: early_data(42) extension does not exist."    \
```
at least this issue reported by the CI.

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]

File: tests/opt-testcases/tls13-misc.sh [hunk 18]
```diff
@@ -493,3 +493,18 @@ run_test "TLS 1.3 m->m: Resumption with ticket flags, psk_all/psk_all." \
...
 
+requires_gnutls_next
+requires_all_configs_enabled MBEDTLS_SSL_EARLY_DATA MBEDTLS_SSL_SESSION_TICKETS     \
+                             MBEDTLS_SSL_SRV_C MBEDTLS_DEBUG_C MBEDTLS_HAVE_TIME    \
+                             MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED \
+                             MBEDTLS_SSL_TLS1_3_COMPATIBILITY_MODE
+requires_any_configs_enabled MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_ENABLED \
+                             MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_EPHEMERAL_ENABLED
+run_test "TLS 1.3 G->m: EarlyData: feature is disabled, fail." \
+         "$P_SRV force_version=tls13 debug_level=4 max_early_data_size=-1 $(get_srv_psk_list)" \
```
Comment (ronald-cron-arm) [inline tests/opt-testcases/tls13-misc.sh:504]:
```suggestion
         "$P_SRV force_version=tls13 debug_level=4 max_early_data_size=-1" \
```
I think the list of PSKs is not necessary here thus it would be better to remove it. No need to re-run the CI for that though. If the second reviewer does not ask for some change, just remove that in a follow-up PR.

**Review by ronald-cron-arm — DISMISSED** [review]
LGTM

File: library/ssl_tls13_server.c [hunk 19]
```diff
@@ -1749,6 +1749,25 @@ static int ssl_tls13_parse_client_hello(mbedtls_ssl_context *ssl,
...
 
+#if defined(MBEDTLS_SSL_EARLY_DATA)
+static void ssl_tls13_update_early_data_status(mbedtls_ssl_context *ssl)
+{
+    mbedtls_ssl_handshake_params *handshake = ssl->handshake;
+
+    if ((handshake->received_extensions &
+         MBEDTLS_SSL_EXT_MASK(EARLY_DATA)) == 0) {
+        MBEDTLS_SSL_DEBUG_MSG(
+            1, ("EarlyData: no early data extension received."));
```
Comment (lpy4105) [inline library/ssl_tls13_server.c:1760]:
```suggestion
        MBEDTLS_SSL_DEBUG_MSG(
            2, ("EarlyData: no early data extension received."));
```
Non-Blocker: The absence of `early_data` extension is not kind of error nor something fatal. So I suggest to use level `2` or `3`.

It's not worth to iterate the PR, we could improve it in a follow-up.

**Review by lpy4105 — APPROVED** [review]
LGTM

**Review by ronald-cron-arm — APPROVED** [review]
LGTM
